File: packages/backend/src/services/tor.ts
import { SocksClient } from 'socks';
import { SocksProxyAgent } from 'socks-proxy-agent';
import { config } from '../config';
import { logger } from '../utils/logger';
import * as fs from 'fs';
import * as path from 'path';

/**
 * TOR service for routing traffic through TOR network
 */
export class TorService {
  private agent: SocksProxyAgent | null = null;
  private hiddenServiceHostname: string | null = null;

  constructor() {
    if (config.tor.enabled) {
      this.initialize();
    }
  }

  private initialize(): void {
    try {
      // Create SOCKS proxy agent for TOR
      const proxyUrl = `socks5://${config.tor.socksHost}:${config.tor.socksPort}`;
      this.agent = new SocksProxyAgent(proxyUrl);
      logger.info('TOR SOCKS proxy agent initialized');

      // Read hidden service hostname if exists
      this.readHiddenServiceHostname();
    } catch (error) {
      logger.error('Failed to initialize TOR service:', error);
    }
  }

  /**
   * Get the SOCKS proxy agent for TOR
   */
  getAgent(): SocksProxyAgent | null {
    return this.agent;
  }

  /**
   * Check if TOR is available
   */
  async checkTorConnection(): Promise<boolean> {
    if (!config.tor.enabled) {
      return false;
    }

    try {
      const info = await SocksClient.createConnection({
        proxy: {
          host: config.tor.socksHost,
          port: config.tor.socksPort,
          type: 5,
        },
        command: 'connect',
        destination: {
          host: 'check.torproject.org',
          port: 80,
        },
        timeout: 10000,
      });

      info.socket.destroy();
      logger.info('TOR connection verified');
      return true;
    } catch (error) {
      logger.error('TOR connection check failed:', error);
      return false;
    }
  }

  /**
   * Read hidden service hostname from TOR directory
   */
  private readHiddenServiceHostname(): void {
    try {
      // Check for both SERVICE1 (backend) and SERVICE2 (web app) .onion addresses
      const service1File = path.join(config.tor.hiddenServiceDir, 'service1', 'hostname');
      const service2File = path.join(config.tor.hiddenServiceDir, 'service2', 'hostname');
      const legacyFile = path.join(config.tor.hiddenServiceDir, 'hostname');

      logger.info(`Checking for .onion addresses...`);

      const checkAndDisplay = () => {
        let backendOnion = '';
        let webOnion = '';

        // Try service1/hostname (multi-service setup)
        if (fs.existsSync(service1File)) {
          backendOnion = fs.readFileSync(service1File, 'utf8').trim();
        }
        // Fallback to legacy location
        else if (fs.existsSync(legacyFile)) {
          backendOnion = fs.readFileSync(legacyFile, 'utf8').trim();
        }

        // Check for web app .onion
        if (fs.existsSync(service2File)) {
          webOnion = fs.readFileSync(service2File, 'utf8').trim();
        }

        if (backendOnion || webOnion) {
          this.hiddenServiceHostname = backendOnion;

          logger.info(`╔════════════════════════════════════════════════════════════════════╗`);
          logger.info(`║                  TOR HIDDEN SERVICES ACTIVE                        ║`);
          logger.info(`╠════════════════════════════════════════════════════════════════════╣`);
          if (backendOnion) {
            logger.info(`║  Backend API:  http://${backendOnion.padEnd(48)} ║`);
          }
          if (webOnion) {
            logger.info(`║  Web App:      http://${webOnion.padEnd(48)} ║`);
          }
          logger.info(`╠════════════════════════════════════════════════════════════════════╣`);
          logger.info(`║  Regular Access:  http://localhost:5173 (web)                     ║`);
          logger.info(`║                   http://localhost:3000 (api)                      ║`);
          logger.info(`╠════════════════════════════════════════════════════════════════════╣`);
          logger.info(`║  Use Tor Browser for .onion addresses for full anonymity          ║`);
          logger.info(`║  Regular browsers work too - messages always encrypted!           ║`);
          logger.info(`╚════════════════════════════════════════════════════════════════════╝`);
          return true;
        }
        return false;
      };

      // Try immediately
      if (checkAndDisplay()) {
        return;
      }

      // If not found, retry periodically
      logger.info('Waiting for Tor to generate .onion addresses...');
      let attempts = 0;
      const maxAttempts = 12;
      const retryInterval = setInterval(() => {
        attempts++;
        if (checkAndDisplay()) {
          clearInterval(retryInterval);
        } else if (attempts >= maxAttempts) {
          logger.warn('Could not find .onion addresses after 60 seconds');
          logger.info('This may be normal if using a different Tor configuration');
          clearInterval(retryInterval);
        }
      }, 5000);
    } catch (error) {
      logger.error('Failed to read hidden service hostname:', error);
    }
  }

  /**
   * Get hidden service .onion address
   */
  getHiddenServiceAddress(): string | null {
    return this.hiddenServiceHostname;
  }

  /**
   * Setup hidden service configuration (for documentation)
   */
  getHiddenServiceConfig(): string {
    return `
# Add to /etc/tor/torrc:

HiddenServiceDir ${config.tor.hiddenServiceDir}
HiddenServicePort 80 127.0.0.1:${config.port}

# Then restart TOR:
# sudo systemctl restart tor

# Your .onion address will be in:
# ${config.tor.hiddenServiceDir}/hostname
    `.trim();
  }

  /**
   * Verify TOR circuit is established
   */
  async verifyCircuit(): Promise<boolean> {
    try {
      // Try to connect through TOR
      const isConnected = await this.checkTorConnection();
      if (isConnected) {
        logger.info('TOR circuit established successfully');
        return true;
      }
      return false;
    } catch (error) {
      logger.error('TOR circuit verification failed:', error);
      return false;
    }
  }

  /**
   * Get TOR connection info
   */
  getConnectionInfo(): {
    enabled: boolean;
    socksHost: string;
    socksPort: number;
    hiddenService: string | null;
  } {
    return {
      enabled: config.tor.enabled,
      socksHost: config.tor.socksHost,
      socksPort: config.tor.socksPort,
      hiddenService: this.hiddenServiceHostname,
    };
  }
}

export const torService = new TorService();

File: rust-backend/src/services/tor.rs
use crate::config::Config;
use crate::error::{AppError, Result};
use std::path::Path;
use tokio::fs;
use tokio_socks::tcp::Socks5Stream;

pub struct TorService {
    config: Config,
}

impl TorService {
    pub fn new(config: Config) -> Self {
        Self { config }
    }

    /// Check TOR connection
    pub async fn check_connection(&self) -> Result<bool> {
        if !self.config.tor_enabled {
            return Ok(false);
        }

        match Socks5Stream::connect(
            format!("{}:{}", self.config.tor_socks_host, self.config.tor_socks_port),
            "check.torproject.org:80",
        )
        .await
        {
            Ok(_) => {
                tracing::info!("TOR connection verified");
                Ok(true)
            }
            Err(e) => {
                tracing::warn!("TOR connection check failed: {}", e);
                Ok(false)
            }
        }
    }

    /// Get hidden service .onion address
    pub async fn get_hidden_service_address(&self) -> Option<String> {
        if !self.config.tor_enabled {
            return None;
        }

        // Try multiple possible locations
        let possible_paths = vec![
            format!("{}/service1/hostname", self.config.tor_hidden_service_dir),
            format!("{}/hostname", self.config.tor_hidden_service_dir),
        ];

        for path in possible_paths {
            if let Ok(content) = fs::read_to_string(&path).await {
                let onion = content.trim().to_string();
                if !onion.is_empty() {
                    tracing::info!("Hidden service found: {}", onion);
                    return Some(onion);
                }
            }
        }

        None
    }

    /// Get connection info
    pub fn get_connection_info(&self) -> ConnectionInfo {
        ConnectionInfo {
            enabled: self.config.tor_enabled,
            socks_host: self.config.tor_socks_host.clone(),
            socks_port: self.config.tor_socks_port,
            hidden_service: None,
        }
    }

    /// Display TOR status on startup
    pub async fn display_status(&self) {
        if !self.config.tor_enabled {
            tracing::info!("TOR: Disabled");
            return;
        }

        let connected = self.check_connection().await.unwrap_or(false);
        let hidden_service = self.get_hidden_service_address().await;

        if connected {
            tracing::info!("╔════════════════════════════════════════════════════════════════════╗");
            tracing::info!("║                  TOR CONNECTION ACTIVE                             ║");
            tracing::info!("╠════════════════════════════════════════════════════════════════════╣");
            tracing::info!("║  SOCKS Proxy: {}:{:<48}║",
                self.config.tor_socks_host,
                self.config.tor_socks_port
            );

            if let Some(onion) = hidden_service {
                tracing::info!("║  Hidden Service: http://{:<43}║", onion);
            }

            tracing::info!("╠════════════════════════════════════════════════════════════════════╣");
            tracing::info!("║  All traffic routed through TOR for maximum privacy               ║");
            tracing::info!("╚════════════════════════════════════════════════════════════════════╝");
        } else {
            tracing::warn!("TOR: Enabled but not connected. Retrying...");
        }
    }
}

#[derive(Debug, serde::Serialize)]
pub struct ConnectionInfo {
    pub enabled: bool,
    pub socks_host: String,
    pub socks_port: u16,
    pub hidden_service: Option<String>,
}
